### ISSUE TITLE
After Win7 install with streamlink-0.1.0.exe, I get "failed to create process."

### ISSUE THREAD

**Issue by jmuff44xv** [streamlink/streamlink#279]
With livestreamer I used to be able to just CD into the directory and run the .exe, but with streamlink I get an error:

C:\Streamlink\bin>streamlink.exe
failed to create process.

(I will add arguments once I get it working.)

Originally, it was in "Program Files (x86)", but I read that the error might be caused by spaces in the path, so I copy-and-pasted the entire Streamlink folder to "C:\" to eliminate spaces.

edit:
The reason I am migrating from livestreamer is I get these errors with livestreamer:

**[cli][error] Unable to open URL: https://www.googleapis.com/youtube/v3/search (4
03 Client Error: Forbidden)**
[cli][info] Waiting for streams, retrying every 450.0 second(s)
**[cli][error] Unable to open URL: https://www.googleapis.com/youtube/v3/search (4
03 Client Error: Forbidden)**

Is there a way to fix this issue with livestreamer?

I hope at least this is fixed with streamlink (if I can solve the "failed to create process" issue).

**Comment by jmuff44xv** [comment]
I just examined the youtube.py file included with streamlink, and except for the "from" statements at the top, it is the same file as what I am using with livestreamer (I manually compiled to .pyc). Therefore, I am not certain that updating to streamlink will fix my original problem. How do I fix this 403 Forbidden Error with YouTube? Does it have something to do with the fact that I am rechecking every 450 seconds? Is that too much? Is there a way to get more info on the nature of the 403 error?

[cli][error] Unable to open URL: https://www.googleapis.com/youtube/v3/search (403 Client Error: Forbidden)

**Comment by beardypig** [comment]
@jmuff44xv that is an issue with the 0.1.0 installer and it will be fixed in the next version. In the mean time you can install the [latest nightly](https://streamlink-builds.s3.amazonaws.com/nightly/windows/streamlink-latest.exe) which should resolve the "failed to create process" error. The error is due to Streamlink using Python 3.5 which requires some extra runtime libraries to be installed, which are not installed when using the 0.1.0 installer. 

As for the 403 error, if it doesn't work with Streamlink we'll try and fix it! Can you give me a YouTube URL that gives that error so I can recreate, debug, and fix your problem?

**Comment by jmuff44xv** [comment]
Thanks! I did not realize it was a known issue. The latest nightly seems to be working, and so far it is not giving me the 403 error I was worried about.

**Comment by beardypig** [comment]
@jmuff44xv No problem, glad you sorted it. Did you have an example URL for the 403 error, might be a bug still :)

**Comment by Vangelis66** [comment]
@jmuff44xv wrote: 

> With livestreamer, **I used to be able to just CD** 
> **into the directory** and run the .exe
> (snip)
> **C:\Streamlink\bin>streamlink.exe**
>  (snip)
> Originally, it was in "Program Files (x86)", 
> but I read that the error might be caused by spaces in the path, 
> so I copy-and-pasted the entire Streamlink folder 
> to "C:" to eliminate spaces.
> (snip)
> **The latest nightly seems to be working**

In the unlikely event you are not already aware, 
both livestreamer and streamlink installers 
add the corresponding executables (more accurately, 
the directories where those executables reside) to PATH, 
so CDing to any directory is not needed

[livestreamer windows binaries](http://docs.livestreamer.io/install.html#windows-binaries) => 
```
Adds Livestreamer to your $PATH (making it possible to use livestreamer directly from the command prompt without specifying its directory)
```
[streamlink windows binaries](https://streamlink.github.io/install.html#windows-binaries) => 
```
Adds Streamlink to your $PATH (making it possible to use streamlink directly from the command prompt without specifying its directory)
```
In the case of livestreamer, if you were using the Nightly zip, 
please kindly disregard my comment, as that was 
not **installed** per se... :smile:

**Comment by karlo2105** [comment]
It works also when you extract nightlies without installing in Program Files, just you have to add this line in streamlink-script.py

`#!"C:\Program Files\streamlink\Python\python.exe"`

**Comment by Vangelis66** [comment]
@karlo2105 wrote: 

> It works also when you extract nightlies without installing in Program Files

Hello karlo; what do you actually mean? 

1. "It works also when you extract nightlies  in Program Files **without installing**"
or 
2. "It works also when you extract nightlies **without installing in Program Files**"

If it's **1.**, what's the actual benefit of doing that? Plus, a user 
needs admin rights to write into %ProgramFiles%; should your code 
be modified for x64 architectures?
`#!"C:\Program Files (x86)\streamlink\Python\python.exe"`
("C:\Program Files (x86)\" being the correct dir to unpack a 32bit app).

If it's **2.**, then portable installations can be found at: 
[streamlink portable](https://github.com/streamlink/streamlink-portable)
(Portable installation creator, by @RosadinTV - BTW, README.md file here 
needs to be modified to include usage directions, as per [previous](https://github.com/streamlink/streamlink-portable/blob/813c2e6/README.md) revision; 
a newbie arriving at this repo isn't sure how to proceed...) 
or 
[streamlink-portable](https://github.com/beardypig/streamlink-portable) => [streamlink-portable-latest-win32](https://s3.amazonaws.com/streamlink-portable/nightly/streamlink-portable-latest-win32.zip)
(experimental - but working currently - already compiled portable installation by @beardypig)

**Comment by karlo2105** [comment]
@Vangelis66 
I use livestreamer without installation as zip file so i did same with streamlink nightlies. I extracted streamlink exe nightlies with 7zip and moved it to program files.
Could you explain me how did you backport streamlink patches and renamed it to livestreamer?

**Comment by Vangelis66** [comment]
> I use livestreamer without installation as zip file

Released version 1.12.2 found [here](https://github.com/chrippa/livestreamer/releases/download/v1.12.2/livestreamer-v1.12.2-win32.zip) or Nightly zip found [here](http://livestreamer-builds.s3.amazonaws.com/livestreamer-latest-win32.zip)?
Either one, it's still a bad idea to extract/unpack anything manually inside %ProgramFiles%

> so i did same with streamlink nightlies. I extracted streamlink exe 
> nightlies with 7zip and moved it to program files.

Bad things to do, IMHO. Streamlink does not come as officially 
endorsed "zip" versions - the nightlies are distributed as NSIS 
installers and are meant to be used as such, i.e. via INSTALLING them!
I will again stress the fact that manually extracting things inside 
Program Files is a faux-pas. If you have to, use your 
%USERPROFILE% dir or similar...

> how did you backport streamlink patches 
> and renamed it to livestreamer?

I am sorry, but I don't get the drift of your question, 
so you'll have to reword it to make it more obvious to me :-(
I have permanently moved on from livestreamer; 
I can assure you that I have not backported 
any streamlink patches to livestreamer, or anywhere 
else for that matter, since I am not a coder :-)

For one last time, if you want to stay on with streamlink, 
either use the installers in the way they are meant to 
(and I believe the only one that warrants support from the devs), 
or (like myself) use one of the two portable flavours 
I linked to in my previous post in this thread... 
Hope I've helped.

**Comment by jmuff44xv** [comment]
@beardypig The command that caused the 403 Forbidden error with livestreamer was this:

**livestreamer https://www.youtube.com/user/jamesnintendonerd/live "720p,480p,360p,240p" --retry-streams 450 -o youtube_live_stream.mp4**

(This is supposed to check for live streams every 450 seconds.)

I also got the same error with a couple of other URLs of this form:
**https://www.youtube.com/user/(ChannelName)/live**

I was worried this was caused by a recent change with YouTube, but so far I have not experienced the issue with streamlink (fingers crossed).

**Comment by jmuff44xv** [comment]
@Vangelis66 Thanks for the info. I think I previously used some kind of portable version of livestreamer, so nothing was added to the PATH variable. Apparently I did not carefully read the note saying "Adds Streamlink to your $PATH" because I incorrectly assumed it was advice telling the user to manually add it to the path (rather than describing what the installer does automatically).

I do like software (especially command line software) that I can install simply by extracting from a .zip or .7z file, and I did try one portable version of streamlink, but I was disappointed that it would not run unless I installed a more recent version of .NET. But I see that you linked to a second portable version that I didn't know about, so I might check that out in the future if I need a portable version.

**Comment by Vangelis66** [comment]
@jmuff44xv wrote: 

> I did try one portable version of streamlink, 
> but I was disappointed that it would not run 
> unless I installed a more recent version of .NET. 
> But I see that you linked to a second portable version 
> that I didn't know about, so I might check that out 
> in the future if I need a portable version.

The requirement for MSVC++2015u3 Redistributable arises from using 
the embedded Windows compile for Python 3.5.2 (WinVistaSP2+); 
installed version as well as both of the portable flavours I linked to 
all employ embedded Python 3.5.2, too!

Since you reported that your latest streamlink nighlty installation 
works as expected, the MSVC++ requirement was taken care of 
by the installer in that case; I am not familiar with the installer code, 
@beardypig is the dev responsible; if the nightly installer installed 
MSVC++2015u3 globally on your machine, then both portable 
flavours would now run OK - if the nightly installer installed it 
in a way that it can only be accessible by the "Programme Files" 
version of streamlink, then I am afraid that to run any portable 
version you'd have to install MSVC++2015u3 manually yourself 
on the machine (prerequisite is [Windows 10 Universal C Runtime](https://www.microsoft.com/en-us/download/details.aspx?id=48234)).

**Comment by beardypig** [comment]
Just so you know, the required DLLs are installed next to the python.exe, so not globally (this is how pynsist does it) :-)

**Comment by beardypig** [comment]
@jmuff44xv sounds like your issue is solved, if you get the error again please open a new issue :-)

**Comment by karlo2105** [comment]
@Vangelis66 

> I am sorry, but I don't get the drift of your question,
so you'll have to reword it to make it more obvious to me :-(

I don't remember name of one active member here who uses the latest streamlink patches for livestreamer backporting for its own purposes.

> For one last time, if you want to stay on with streamlink,
either use the installers in the way they are meant to
(and I believe the only one that warrants support from the devs),
or (like myself) use one of the two portable flavours
I linked to in my previous post in this thread...
Hope I've helped.

Don't worry I just experienced with streamlink. 
I already have Python 2.7.11 installed so I would like to use this version of Python instead of Python 3.5.2 included with Streamlink.

**Comment by Vangelis66** [comment]
@beardypig wrote: 

> the required DLLs are installed next to the python.exe, so not globally

Just so you know, **Windows6.0-KB2999226-x86.msu** update ( Windows 10 Universal C Runtime 
for Windows Vista SP2 32bit) is pre-installed in my system via Windows Update. 
However, I have not (that I know of) manually installed the MSVC++2015 redistributable 
nor does it appear under "Programs and features" next to previous runtimes: 

![runtimes](https://cloud.githubusercontent.com/assets/9669492/21077324/ae6a788a-bf50-11e6-876d-513010de4983.jpg)

I have not witnessed the nightly installer place any additional required DLLs next to the python.exe: 

![pythonfolder](https://cloud.githubusercontent.com/assets/9669492/21077353/70610832-bf51-11e6-9946-04c86cc57912.jpg)

File **vcruntime140.dll** (v14.0.23918.0) is the one dll pre-bundled with [python-3.5.2-embed-win32.zip](https://www.python.org/ftp/python/3.5.2/python-3.5.2-embed-win32.zip) 
which, if I'm not mistaken, is the embedded Py 3.5.2 version used in the installer. 
This dll is also included by default in both portable flavours of streamlink (see my previous post), 
so under my setup I can run installed/portable streamlink without having MSVC++2015 
itself installed. One possible caveat is that the bundled vcruntime140.dll (without which Py 3.5.2 
fails to run) is a tad outdated, as the latest file version (installed globally by latest redistributable) 
is v14.0.24215.1.
Have I failed to properly understand things?

@karlo2105 

> I already have Python 2.7.11 installed

I think latest 2.7.x version is **2.7.12**

> instead of Python 3.5.2 included with Streamlink.

The latter being an embedded (secluded) version, 
it is unlikely to interact with your installed 2.7.x version...

**Comment by beardypig** [comment]
@Vangelis66you got it. If you want the specifics, it installs the UCRT DLLs only if you need them, ie you don't have KB2999226 installed. It checks for the presence of ucrtbase.dll to detect KB2999226. The version of vcruntime140.dll that is bundled with the embedded Python 3.5.2 distribution is the corresponding version for the version of MSVC that was used to compile Python 3.5.2.

**Comment by Vangelis66** [comment]
@beardypig wrote: 

> you got it. If you want the specifics, it installs the UCRT DLLs 
> only if you need them, ie you don't have KB2999226 installed

As you might have gathered by now, I have an inquisitive nature, 
but I also know when to draw a line and stop digging deeper; 
this is one of those cases :smiley:

> It checks for the presence of ucrtbase.dll to detect KB2999226.

Just for the record, I manually searched for that specific file, 
it's to be found inside  
`C:\Windows\System32\`
and the file version reads **10.0.10586.9**; "Date created" 
coincides with the date KB2999226 was installed, so all add up :+1:

**Comment by jmuff44xv** [comment]
@Vangelis66 

> The requirement for MSVC++2015u3 Redistributable arises from using
the embedded Windows compile for Python 3.5.2

Actually I believe the error message I received when I tried this [portable release](https://github.com/streamlink/streamlink-portable/tree/master/Streamlink%20for%20Windows%20%28Compiled%29) was about the .NET Framework rather than the Visual C++ runtime.

**Comment by Vangelis66** [comment]
@jmuff44xv wrote: 

> the error message I received when I tried this [portable release](https://github.com/streamlink/streamlink-portable/tree/master/Streamlink%20for%20Windows%20%28Compiled%29) 
> was about the .NET Framework rather than the Visual C++ runtime.

What I wrote stands true even for that portable iteration, 
since it, too, employs Python 3.5.2 embedded for Windows 
(file [python-3.5.2-embed-win32.zip](https://www.python.org/ftp/python/3.5.2/python-3.5.2-embed-win32.zip)). 

The additional requirement for .NET Framework 4.5+ stems from the fact 
the "**Streamlink for Windows.exe**" executable needs that minimum 
.NET FW version installed in the host machine to be able to run.
